### PR TITLE
Feature | Naive signup capabilities implementation

### PR DESCRIPTION
**Pull request by EstebanBorai**
A naive version of the `signup` service is implemented,
providing the capability of registering users in the platform.

Its important to note that theres a lot of work to be done in
this feature, including:

- e2e testing
- Validations on incoming payload
- Code cleaning
- Error handling
- Configure SQLx for offline mode to make workflows run

And perhaps theres more points I will encounter for the moment
I work on those missing items.

I think is important to merge this version as this establishes a
project scaffold for all of us to contribute/expand this project.

## Example

In order to register a new user you must send an HTTP request
which looks like the following:

```
POST /signup HTTP/1.1
Content-Type: application/json; charset=utf-8
Host: 0.0.0.0:7878
Connection: close
Content-Length: 101

{"email":"eborai@hilow.com","username":"eborai","name":"Esteban","surname":"Borai","password":"1234"}
```

Note: This PR depends on #18 and #22

### CODE REVIEW

File: .env.development
Comment (morenol) [inline .env.development:1]: If we remove this file, then the program will panic in debug mode, right?

Comment (morenol) [reply .env.development:1]: nm, I just saw that you changed that also in the main.rs file

Comment (morenol) [inline .env.development:1]: I think that we should keep this file. It was working by default, now we need to create a .env file to make this works.

Comment (EstebanBorai) [reply .env.development:1]: Thing is, not everybody is going to use the same environment, some people may want to use different values for such environment variables.

The most common pattern is to have a `.env.sample` which provides predefined values, but the project will consume a copy of that file under the name `.env` which is ignored by Git and every collaborator will keep a copy of that file locally.

File: src/domain/services/mod.rs
Comment (EstebanBorai) [inline src/domain/services/mod.rs:31]: This should be implemented in a different manner, using a Service to wrap each repository so we can have control on data manipulation and then inserting this data on the database.

Currently is being delivered very straightforward to open the project to other collaborators. I will land a Refactor PR with improvements on this approach.

File: src/domain/services/secret.rs
Comment (EstebanBorai) [inline src/domain/services/secret.rs:37]: I think is a good idea to keep best practices here and investigate if we can provide a custom secret to this `Config` for hashing/salting.

File: src/domain/services/user.rs
Comment (EstebanBorai) [inline src/domain/services/user.rs:63]: As is clear here, everything is being done pretty straightforward, we are missing validations on these details which is not okay. This is improved later as well.

File: src/domain/services/user.rs
Comment (EstebanBorai) [inline src/domain/services/user.rs:69]: Here I plan to have some password validation to make sure password is strong enough before moving any further in registering process

File: src/handlers/signup.rs
Comment (EstebanBorai) [inline src/handlers/signup.rs:24]: These clones are awful but we can't move fields out of `payload` so.. I will check how can we improve this I'm not 100% sure if this is a common pattern for `Json` type.

**Review by EstebanBorai — COMMENTED** [review]
Added comments on strange stuff

File: src/domain/repositories/profile.rs
Comment (morenol) [reply src/domain/repositories/profile.rs:28]: It looks like query_file is crate only, sqlx-cli does not have that feature enabled.

https://docs.rs/sqlx/0.4.1/sqlx/macro.query_file.html
https://github.com/launchbadge/sqlx/blob/master/sqlx-cli/Cargo.toml

Comment (morenol) [reply src/domain/repositories/profile.rs:28]: I tried to put all this changes on top of the other PR that I have. Then I ran cargo sqlx migrate run.

cargo sqlx prepare -- --bin main fails with:
```rust
  --> src/domain/repositories/profile.rs:28:21
   |
28 |         match sqlx::query_file!("sql/profiles/insert.sql", name, surname, user_id)
   |                     ^^^^^^^^^^ could not find `query_file` in `sqlx`

```

Comment (EstebanBorai) [reply src/domain/repositories/profile.rs:28]: Pretty strange, make sure you run `cargo check` and see if ir compiles. I was able to run this locally with no major issues.

Im unable to check at the moment but will be checking in the next hours.

**Comment by EstebanBorai** [comment]
@morenol if this PR results as a stopper for other PRs authored by you, feel free to go ahead with #18 and #22 and I will rebase against `main` when those PRs are landed.

At the moment Im not able to fix those but will do later today!

**Comment by rafacouto** [comment]
I suggest to move the URI under /users since operations are more related to user than session.

File: src/domain/repositories/profile.rs
Comment (morenol) [reply src/domain/repositories/profile.rs:28]: I tried again, and it worked so just ignore that comment. I am Ok to merge this after 22 is merged and a rebase (as long as the CI is green!)

**Review by morenol — APPROVED** [review]

File: src/bin/main.rs
Comment (morenol) [inline src/bin/main.rs:19]: ?
```suggestion
        ?;
```

Comment (EstebanBorai) [reply src/bin/main.rs:19]: Nice catch!

**Comment by EstebanBorai** [comment]
> I suggest to move the URI under /users (/users/signup or /users/register) since operations are more related to user than session.

I don't th

> I suggest to move the URI under /users (/users/signup or /users/register) since operations are more related to user than session.

Thing is that neither `login` or `signup` consume a User resource. Neither of these endpoints relates to a User. When interacting with authentication you provide credentials such as email or username and password and you don't expect the User resource to be in the response from the server but a Token which is then used to gather User relevant details required by the client.

If you check the documentation: https://github.com/rust-lang-ve/hilow#user-management, you will notice that theres an endpoint for the `User` resource which retrieves the user for the client, this requires the token you gather when consuming either this or the login endpoint.

File: src/handlers/signup.rs
Comment (EstebanBorai) [reply src/handlers/signup.rs]: This was easily fixed using `RegisterUserDto` straight and `into_inner`!

File: src/domain/services/user.rs
Comment (EstebanBorai) [reply src/domain/services/user.rs:69]: Passwords and Emails are going to be validated based on a configuration. Thus we are leaving this as `true` while its developed

**Comment by rafacouto** [comment]
> > I suggest to move the URI under /users (/users/signup or /users/register) since operations are more related to user than session.
> 
> Thing is that neither `login` or `signup` consume a User resource. Neither of these endpoints relates to a User. When interacting with authentication you provide credentials such as email or username and password and you don't expect the User resource to be in the response from the server but a Token which is then used to gather User relevant details required by the client.
> 
> If you check the documentation: https://github.com/rust-lang-ve/hilow#user-management, you will notice that theres an endpoint for the `User` resource which retrieves the user for the client, this requires the token you gather when consuming either this or the login endpoint.

_login_ and _logout_ are clearly related to session (not user) so I'd move them to /session path. However register is a method for registering users. You are mixing URI path and authentication concepts and they are not linked in modern APIs. The goal here should be to understand method domains. Anyway, I accept the PR!

**Review by rafacouto — APPROVED** [review]

**Comment by EstebanBorai** [comment]
> > > I suggest to move the URI under /users (/users/signup or /users/register) since operations are more related to user than session.
> > 
> > Thing is that neither `login` or `signup` consume a User resource. Neither of these endpoints relates to a User. When interacting with authentication you provide credentials such as email or username and password and you don't expect the User resource to be in the response from the server but a Token which is then used to gather User relevant details required by the client.
> > 
> > If you check the documentation: https://github.com/rust-lang-ve/hilow#user-management, you will notice that theres an endpoint for the `User` resource which retrieves the user for the client, this requires the token you gather when consuming either this or the login endpoint.
> 
> _login_ and _logout_ are clearly related to session (not user) so I'd move them to /session path. However register is a method for registering users. You are mixing URI path and authentication concepts and they are not linked in modern APIs. The goal here should be to understand method domains. Anyway, I accept the PR!

The definition of routes was implemented iterations before.

We could address this in a Discussion perhaps?

This PR implements features beyond routing and in case routes have to be updated we could land another PR in the future introducing such change.

**Pull request merged**
